From D.Barow@option.com  Wed Aug 20 13:26:56 2008
From: Denis Joseph Barrow <D.Barow@option.com>
Date: Wed, 20 Aug 2008 16:34:01 +0200
Subject: HSO: fix hso_serial_close refcounting
To: Linux USB kernel mailing list <linux-usb@vger.kernel.org>, Greg Kroah-Hartman <gregkh@suse.de>
Message-ID: <48AC2B59.1090101@option.com>


The hso: fix refcounting on the ttyHSx devices signed off by Olivier
Blin was better than what was there but not completely correct. The
kref_put has to be moved below the mutex_unlock otherwise the memory
used by the mutex could be freed before it is used.

Signed-off-by: Denis Joseph Barrow <D.Barow@option.com>
Cc: Olivier Blin <blino@mandriva.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/net/usb/hso.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/usb/hso.c
+++ b/drivers/net/usb/hso.c
@@ -1290,7 +1290,6 @@ static void hso_serial_close(struct tty_
 	/* reset the rts and dtr */
 	/* do the actual close */
 	serial->open_count--;
-	kref_put(&serial->parent->ref, hso_serial_ref_free);
 	if (serial->open_count <= 0) {
 		serial->open_count = 0;
 		if (serial->tty) {
@@ -1307,6 +1306,7 @@ static void hso_serial_close(struct tty_
 		usb_autopm_put_interface(serial->parent->interface);
 
 	mutex_unlock(&serial->parent->mutex);
+	kref_put(&serial->parent->ref, hso_serial_ref_free);
 }
 
 /* close the requested serial port */
